### PR TITLE
[SPARK-31563][SQL] Fix failure of InSet.sql for collections of Catalyst's internal types

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to fix the `InSet.sql` method for the cases when input collection contains values of internal Catalyst's types, for instance `UTF8String`. Elements of the input set `hset` are converted to Scala types, and wrapped by `Literal` to properly form SQL view of the input collection.

### Why are the changes needed?
The changes fixed the bug in `InSet.sql` that makes wrong assumption about types of collection elements. See more details in SPARK-31563.

### Does this PR introduce any user-facing change?
Highly likely, not.

### How was this patch tested?
Added a test to `ColumnExpressionSuite`

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
Looks plausible. This needs to go to 2.4 too right?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for splitting your existing PR, @MaxGekk .
Yes, @srowen . This is a fix for all 2.x. So, I asked to split the original PR.
Merged to master/3.0/2.4.

**Comment by dongjoon-hyun** [comment]
cc @holdenk since she is the release manager for 2.4.6.

**Comment by holdenk** [comment]
Thanks everyone :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 1]
```diff
@@ -519,7 +520,9 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
 
   override def sql: String = {
     val valueSQL = child.sql
-    val listSQL = hset.toSeq.map(Literal(_).sql).mkString(", ")
+    val listSQL = hset.toSeq
+      .map(elem => Literal(convertToScala(elem, child.dataType)).sql)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:524]: this converts the internal value to external value, and then `Literal.apply` converts external value to internal value.

Can we just do `Literal(elem, child.dataType).sql`?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:524]: @cloud-fan No, Literal fails on UTF8String value, see https://github.com/apache/spark/pull/28328#discussion_r414566719

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:524]: Another problem is https://github.com/apache/spark/pull/28328#discussion_r415256811

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:524]: So, the problem is Literal requires external types but elem has internal Catalyst's type

**Pull request closed without merging**